### PR TITLE
Add mtopi CSR

### PR DESCRIPTION
**Pull request by KushalMeghani1644**
This PR adds support for the `mtopi` (Machine Top Priority Interrupt) CSR, which is part of the RISC-V Advanced Interrupt Architecture (AIA) extension.

**Changes:**
- Add `mtopi.rs` register implementation at address 0x7C0
- Export mtopi module in `register.rs`
- Read-only CSR with proper bit field definitions:
  - `iid` (Interrupt ID): bits 16-27 
  - `ipid` (Interrupt Priority ID): bits 0-7
- Convenience methods for interrupt detection and priority handling

**Features:**
- Memory-safe read-only access
- Comprehensive documentation with usage examples
- Full test coverage
- Follows existing CSR implementation patterns

**Testing:**
- All existing tests continue to pass
- New tests cover bit field extraction and edge cases
- Documentation examples compile successfully

**AI Assistance:**
Parts of the explanation and understanding of the mtopi CSR were assisted using an AI tool to summarize its behavior in the context of the RISC-V Advanced Interrupt Architecture (AIA). All code and implementation decisions were written and verified manually. NO CODES WERE GENERATED WITH THE HELP OF AI.

### CODE REVIEW

**Review by rmsyn — CHANGES_REQUESTED** [review]
Thanks for your contribution!

> AI Assistance:
Parts of the explanation and understanding of the mtopi CSR were assisted using an AI tool to summarize its behavior in the context of the RISC-V Advanced Interrupt Architecture (AIA). All code and implementation decisions were written and verified manually. NO CODES WERE GENERATED WITH THE HELP OF AI.

Thank you for explicitly stating your use of AI tool assistance. AFAIK, we currently do not have a policy on AI contributions. IMO, we should ban AI contributions, even for "just documentation/explanation". Such contributions have deep issues with plagiarism, and we cannot possibly verify the provenance of of the training data.

"Manually" rewriting suggestions from AI is not the same as manually implementing the changes. You're already contradicting yourself by saying that AI only helped by providing explanations/documentation, and that you manually rewrote *code* suggested by AI.

@romancardenas what are your thoughts on adding a section on AI to our `CODE_OF_CONDUCT.md`, or creating a `CONTRIBUTING.md`? After your recent back and forth on @KushalMeghani1644's contributions, how do you feel about their usage of AI?

**Comment by rmsyn** [comment]
> Hey @rmsyn thanks for putting time to my PR! I feel there is maybe a little misunderstanding upon the "AI DISCLOSURE" part, I have clearly stated in the PR desc that:
> 
> "Parts of the explanation and understanding of the mtopi CSR were assisted using an AI tool to summarize its behavior in the context of the RISC-V Advanced Interrupt Architecture (AIA). All code and implementation decisions were written and verified manually. NO CODES WERE GENERATED WITH THE HELP OF AI."
> 
> See the "parts" word! I meant that only a few little queries and all were asked from AI to understand how Mtopi actually works since I haven't studied much on this CSR, and the codes WERE NOT generated by AI, I had indeed went through the official docs of https://tools.cloudbear.ru/docs/riscv-aia-1.0-20230630.pdf to understand mtopi well. And also this line by you, ""Manually" rewriting suggestions from AI is not the same as manually implementing the changes. You're already contradicting yourself by saying that AI only helped by providing explanations/documentation, and that you manually rewrote code suggested by AI." I hadn't not _manually rewritter_ what AI gave me, I stated in the "AI DISCLOSURE" section in capital words that, "NO CODES WERE GENERATED WITH THE HELP OF AI."
> 
> Edit: Also just to be transparent, my previous PRs that were merged had 0 use of AI in the codes! Those PRs were purely made by my brain + hands :D
> 
> Thank you :)

The official link to the AIA specification (from the [RISC-V Atlassian repo](https://riscv.atlassian.net/wiki/spaces/HOME/pages/16154732/Ratified+Extensions)) is a Google Docs link [RISC-V AIA](https://drive.google.com/file/d/1joBC2hWGEHJL4tFabjcqMpRqQNkqJ5WR/view?usp=drive_link)], maybe the cloudbear.ru link you posted is a mirror of that document?

Regarding your previous contributions, that's fine if you indeed contributed all those changes yourself. I was asking for @romancardenas to give feedback on allowing your AI contirbutions after the experience of reviewing your previous contribution.

**Comment by rmsyn** [comment]
> Hey @rmsynI've updated the mtopi tests to use test helper macros as requested. Created a `test_ro_csr_field!` macro that follows the same pattern as `test_csr_field!` from macros.rs but adapted for read-only CSRs. The macro uses identical bit manipulation logic and replaces all the manual assertions. Thanks for the guidance!

Please do not add a new macro, instead just add a new macro arm to the existing `test_csr_field!` macro.

**Comment by KushalMeghani1644** [comment]
Hey! @rmsyn (sorry for pinging if you were busy) I've added a new arm to the existing test_csr_field! macro instead of creating a separate one. The macro now supports read-only CSR fields with the pattern [start:literal, end:literal] and mtopi tests now use the standard macro. Thanks for the clarification!

**Comment by rmsyn** [comment]
> Hey! @rmsyn (sorry for pinging if you were busy) I've added a new arm to the existing test_csr_field! macro instead of creating a separate one. The macro now supports read-only CSR fields with the pattern [start:literal, end:literal] and mtopi tests now use the standard macro. Thanks for the clarification!

No worries, and thanks for making the changes. I have a couple other suggestions, otherwise LGTM outside the question of AI contributions.

File: riscv/src/register/mtopi.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    /// Returns the interrupt identifier
+    ///
+    /// A value of 0 indicates no interrupt is pending. Non-zero values identify
+    /// specific interrupt sources as defined by the interrupt controller configuration.
+    #[inline]
+    pub fn interrupt_id(&self) -> usize {
+        self.iid()
+    }
```
Comment (rmsyn) [inline riscv/src/register/mtopi.rs:72]: I don't know if we need these convenience functions, and I don't think we use this pattern anywhere else in the crate. `has_interrupt` is fine, because it clarifies the semantics of the `iid` field.

**Review by rmsyn — CHANGES_REQUESTED** [review]
> Hey! @rmsyn (sorry for pinging if you were busy) I've added a new arm to the existing test_csr_field! macro instead of creating a separate one. The macro now supports read-only CSR fields with the pattern [start:literal, end:literal] and mtopi tests now use the standard macro. Thanks for the clarification!

No worries, and thanks for making the changes. I have a couple other suggestions, otherwise LGTM outside the question of AI contributions.

**Comment by KushalMeghani1644** [comment]
Hey @rmsyn and @romancardenas sorry for disturbing but can I please get any updates on what do I have to do with this PR next?

File: riscv/src/register/mtopi.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+//!     println!("Highest priority interrupt: ID={}, Priority={}", interrupt_id, priority);
+//! } else {
+//!     println!("No interrupts pending");
+//! }
+//! ```
+
+read_only_csr! {
+    /// Machine Top Priority Interrupt Register
+    Mtopi: 0x7C0,
+    mask: usize::MAX,
```
[the remark this answers is not in the document]
Comment (KushalMeghani1644) [reply riscv/src/register/mtopi.rs]: I have fixed the bits to 0 to 27, thanks for pointing it out.

File: riscv/src/register/mtopi.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    iid: [16:27],
+}
+
+read_only_csr_field! {
+    Mtopi,
+    /// Interrupt Priority ID (bits 0..7)
+    ///
+    /// Represents the priority level of the pending interrupt.
+    /// Higher numerical values indicate higher priority interrupts.
+    ipid: [0:7],
```
[the remark this answers is not in the document]
Comment (KushalMeghani1644) [reply riscv/src/register/mtopi.rs]: I am sorry for such silly mistakes, I had mixed iprio, and iid resulting to `ipid` :) I have also fixed this. The documentation I followed is this [tools.cloudbear.ru/docs/riscv-aia-1.0-20230630.pdf](https://tools.cloudbear.ru/docs/riscv-aia-1.0-20230630.pdf) and https://drive.google.com/file/d/1joBC2hWGEHJL4tFabjcqMpRqQNkqJ5WR/view?pli=1

**Comment by KushalMeghani1644** [comment]
A unrelated change @romancardenas (maybe) but I have fixed the spelling in aclint.rs, I hope you are fine with that :) if you think its unrelated to this PR I can also remove the commits.

File: riscv/src/register/macros.rs [hunk 4]
```diff
@@ -1041,6 +1041,19 @@ macro_rules! test_csr_field {
         }
     }};
 
+    // test a multi-bit bitfield for read-only CSR (must come before enum pattern)
+    ($reg:ident, $field:ident: [$start:literal, $end:literal]) => {{
```
Comment (romancardenas) [inline riscv/src/register/macros.rs]: Why do you need to add new test variants? Can't you use current patterns?

**Comment by romancardenas** [comment]
Looks like the new version of syn bumped the MSRV to 1.68... 

I was thinking on bumping MSRV to 1.81 with riscv-types, so it shouldn't be a big deal. Let put this PR in standby until I do all the required work.

File: riscv/src/register/macros.rs [hunk 5]
```diff
@@ -1041,6 +1041,18 @@ macro_rules! test_csr_field {
...
+    ($reg:ident, $field:ident: [$start:expr, $end:expr]) => {{
+        let bits = $reg.bits();
+        let exp_val = $crate::bits::bf_extract(bits, $start, $end - $start + 1);
+        assert_eq!($reg.$field(), exp_val);
+    }};
+
+    // test a multi-bit bitfield for read-only CSR with expected value
+    ($reg:ident, $field:ident: [$start:expr, $end:expr], $expected:expr) => {{
+        assert_eq!($reg.$field(), $expected);
+    }};
```
[the remark this answers is not in the document]
Comment (romancardenas) [reply riscv/src/register/macros.rs]: Only one new test branch that does all the checks

File: riscv/src/register/mtopi.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    use super::*;
+
+    #[test]
+    fn test_mtopi_fields() {
+        let mtopi = Mtopi::from_bits(0);
+
+        test_csr_field!(mtopi, iid: [16, 27], 0x0);
+        test_csr_field!(mtopi, iprio: [0, 7], 0x0);
+
+        let mtopi = Mtopi::from_bits((11 << 16) | 5);
```
Comment (romancardenas) [inline riscv/src/register/mtopi.rs]: For clarity, so the numbers coincide with the next checks
```suggestion
        let mtopi = Mtopi::from_bits((11 << 0xB) | 0x5);
```

Comment (romancardenas) [reply riscv/src/register/mtopi.rs]: My bad, I added a bug to your tests! In any case, it served me to notice that tests are not currently being executed in the CI.

**Comment by KushalMeghani1644** [comment]
I have messed up something here.... letme convert this into a draft, when its ready I'll mark it ready for review

**Comment by romancardenas** [comment]
There are too many changes in your PR that are not required. As reviewing this would take too long, I will close this PR. You can create a new PR from scratch, starting from the current master branch, with a clean commit history.
